### ISSUE TITLE
Not able to reproduce the published results of GAT on Cora dataset

### ISSUE THREAD

**Issue by ekagra-ranjan** [rusty1s/pytorch_geometric#338]
## ❓ Questions & Help

Hi! 

I am trying to reproduce the results of GAT on Cora dataset. I took the hyperparameters (learning rate, patience for early stopping, L2 norm, dropout, number of heads and hidden features) from the paper and used the GATConv provided by PyG. I am able to get only ~79% as an average of 10 runs. The reproduced results are not even better than a GCN (~80.4%) . Did anyone achieve the published result of 81.4% on Cora using GAT?

Paper: https://arxiv.org/abs/1710.10903

Here is my code used for reproducing the results:

Model:
```
import torch
import torch.nn.functional as F
from torch_geometric.nn import GCNConv, SAGEConv, GATConv

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        hidden_state = 8
        self.conv1 = GATConv(dataset.num_features, hidden_state, heads=8, dropout=0.6)
        self.conv2 = GATConv(hidden_state*8, dataset.num_classes, heads=1, dropout=0.6)


    def forward(self, data):
        x, edge_index = data.x, data.edge_index

        x = F.dropout(x, p=0.6, training=self.training)
        x = self.conv1(x, edge_index)
        x = F.elu(x)
        x = F.dropout(x, p=0.6, training=self.training)
        x = self.conv2(x, edge_index)

        return F.log_softmax(x, dim=1)
```

Code for training and benchmarking
```
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("device: ", device)
    
val_acc=0.0
test_acc=0.0

for i in range(10):
    
    model = Net().to(device)
    data = dataset[0].to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)

    model_best = Net().to(device)

    count=0
    patience=100
    lossMin = 1e10
    
    print(i, end=' ')
    model.train()
    for epoch in range(800):
        optimizer.zero_grad()
        out = model(data)
        loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])
        loss.backward()    
        optimizer.step()

        #Validation
        model.eval()
        _, pred = model(data).max(dim=1)
        correct = float (pred[data.val_mask].eq(data.y[data.val_mask]).sum().item())
        val_loss = F.nll_loss(out[data.val_mask], data.y[data.val_mask])
        
        #early stopping
        if (val_loss<=lossMin):
            lossMin = val_loss
            count = 0
            model_best.load_state_dict(model.state_dict())
            val_acc_best = (correct / data.val_mask.sum().item())
        else:
            count+=1
            if(count>patience):
                print("patience lost", end=' ')
                break
                
    val_acc += val_acc_best
    
    #Testing
    model_best.eval()
    _, pred = model_best(data).max(dim=1)
    correct = float (pred[data.test_mask].eq(data.y[data.test_mask]).sum().item())
    test_acc += (correct / data.test_mask.sum().item())
    print("correct ", correct)


print("Avg Val acc:", val_acc/10)
print("Avg Test acc:", test_acc/10)
```

Output
```
device:  cuda
0 patience lost correct  774.0
1 patience lost correct  778.0
2 patience lost correct  792.0
3 patience lost correct  795.0
4 patience lost correct  790.0
5 patience lost correct  785.0
6 patience lost correct  801.0
7 patience lost correct  790.0
8 patience lost correct  778.0
9 patience lost correct  796.0
Avg Val acc: 0.7767999999999999
Avg Test acc: 0.7879
```

**Comment by rusty1s** [comment]
You only set `model.train()` once before training start. Hence, you train without dropout after the first epoch. This increases the performance of your model.

**Comment by ekagra-ranjan** [comment]
Thanks a lot rusty1s! I also found your bechmarks quite useful. I will be closing this Issue then

**Comment by xuehui1991** [comment]
@rusty1s  hello. I rerun the code in `benchmark/citation`, the result as following:
![image](https://user-images.githubusercontent.com/6280746/69114804-63f22180-0ac1-11ea-9f44-4a121af7db67.png)

Seems still has some difference with the paper on Cora?

Thank you~

**Comment by ekagra-ranjan** [comment]
If the difference is small, it is due to the fact that GAT was originally implemented on Tensorflow: [Original Repo](https://github.com/PetarV-/GAT). Results in implementations in Pytorch and Tensorflow can differ by small margins.
